# 12.6 Data Cleaning Challenge

In [887]:
import numpy as np
import pandas as pd


To complete this challenge, determine the five most common journals and the total articles for each. 

Next, calculate the mean, median, and standard deviation of the open-access cost per article for each journal.

You will need to do considerable data cleaning in order to extract accurate estimates. You may may want to look into data encoding methods if you get stuck. For a real bonus round, identify the open access prices paid by subject area.

Remember not to modify the data directly. Instead, write a cleaning script that will load the raw data and whip it into shape. Jupyter notebooks are a great format for this. Keep a record of your decisions: well-commented code is a must for recording your data cleaning decision-making progress. 

In [888]:

#import CSV file and set to dataframe
df = pd.read_csv("WELLCOME_APCspend2013_forThinkful.csv", encoding='latin1')

#look at the top of the dataframe
#df.head()
#df['Journal title'].unique()

In [889]:
#'JOURNAL OF BIOLOGICAL CHEMISTRY','The Journal of Biological Chemistry',
#'BMC Genomics.' ,''BMC Genomics'
#'British Journal for the History of Science',
#'British Journal of the History of Science',


In [890]:
#count the number of unique values for each of 
#the columns  and see which ones are the most prevalent
count=df.groupby('Journal title').nunique()
count = count.sort_values('Article title')
count.describe()
#count.tail(60)

PMID/PMCID   Publisher  Journal title  Article title  \
count  984.000000  984.000000          984.0     984.000000   
mean     1.958333    1.244919            1.0       2.159553   
std      4.308700    0.648948            0.0       4.375083   
min      0.000000    1.000000            1.0       1.000000   
25%      1.000000    1.000000            1.0       1.000000   
50%      1.000000    1.000000            1.0       1.000000   
75%      2.000000    1.000000            1.0       2.000000   
max     91.000000    7.000000            1.0      91.000000   

       COST (£) charged to Wellcome (inc VAT when charged)  
count                                         984.000000    
mean                                            1.878049    
std                                             3.455187    
min                                             1.000000    
25%                                             1.000000    
50%                                             1.000000    
75%                                             2.000000    
max                                            68.000000

In [891]:
# create a dictionary with typos to be corrected
#that need to be replaced in the journal title column
typos = {' j ':'Journal',
         ' J ':'Journal',
         'J ':'Journal',
         'JAIDS': '',
         'Syndroms':'Syndromes',
         'Archiv':'Archive',
         'Sex Transm Infect.':'Sexually Transmitted Infections',
         'Agfents':'Agents',
         ' Biol ': ' Biological ',
         'Infect.':'Infections',
         'Sci.':'Science',
         'bioohysica': 'biophysica',
         'Int ':'International ',
         'Int.':'International',
         'for parasitology':'of parasitology',
         'for clinical nutrition':'of clinical nutrition',
         'of the history':'for the history',
         'internationalrnational': 'international'    
        }

for key in typos:
    df['Journal title'] = df['Journal title'].str.replace(key,typos[key])



In [892]:
# create a dictionary with string elements
#that need to be replaced in the journal title column
replacments = {'_'  : ' ',
              '&'  : 'and',
               '  ':' ',
              '()':''
              }

#iterate through the dataframe and replace any strings found with strings from the dictionary 
for key in replacments:
    df['Journal title'] = df['Journal title'].str.replace(key,replacments[key])
    

In [893]:
#show unique journal titles to search for errors 
#df['Journal title'].unique()

In [894]:
#create list of improper titles for PLoS Neglected Tropical Diseases
PloseNTDList = ['PLoS Neglected Tropical Diseases',
               'PLOS Neglected Tropical Diseases',
               'PLoS Negected Tropical Diseases',
               'Neglected Tropical Disease']
#correct titles from list to consolodate articles 
for name in PloseNTDList:
    df['Journal title']= df['Journal title'].replace([name],'PLoS Neglected Tropical Diseases')


In [895]:
NARlist = ['Nucleic Acids Research','Nucleic Acids Research\n',
           'Nucleic Acids Research ','NUCLEIC ACIDS RESEARCH',
           'Nucleic Acids Research','Nucleic Acid Research'
           'NUCLEIC ACIDS RESEARCH','Nucleic Acids Research ',
           'Nucleic Acids Research\n','Nucleic acid research']
#correct titles from list to consolodate articles 

for name in NARlist:
    df['Journal title']= df['Journal title'].replace([name],'Nucleic Acids Research')


In [896]:

JnlBioChemlist =['Journal of Biological Chemistry','JOURNAL OF BIOLOGICAL CHEMISTRY',
                'The Journal of Biological Chemistry','Journal of biological chemistry',
                'The journal of Biological Chemistry','The journal of Biological Chemistry',
                'Jnl Biological Chemistry','Journal of Biol Chem','Journal Biological Chemistry',
                'Journal of Biological Chem']

#correct titles from list to consolodate articles 
for name in JnlBioChemlist:
    df['Journal title']= df['Journal title'].replace([name],'Journal of Biological Chemistry')


In [897]:
#Neuroimage
Neuroimagelist =['NeuroImage',
       'Neuroimage: Clinical', 'Neuroimage: clinical',
       'NeuroImage: Clinical', 'NeuroImage: Clinical ', 'Neurolmage']

#correct titles from list to consolodate articles 
for name in Neuroimagelist:
    df['Journal title']= df['Journal title'].replace([name],'Neuroimage')


In [898]:
ploslist= ['PLoS 1', 'PLoS ONE','PloS One',
           'PLOS','Plos','PLoS','PLOS  ONE',
           'PLoS 1','PloS ONE','PLosONE',
           'PLOS One','PLOS ONE','PLoS One',
          'Plos ONE','PLoSONE','Plos ONE',
           'Plos one','PLoS One','Plos One']

#correct titles from list to consolodate articles 
for name in ploslist:
    df['Journal title']= df['Journal title'].replace([name],'PLoS One')


In [899]:
pnaslist= ['Proceedings of the national academy of sciencences (pnas)',
           'Proceedings of the national academy of sciencences',
           'Proceedings of the national academy of sciencences of the usa', 
           'Pnas usa',
           'Proceedings of the national academy of sciencences of the united states of america',
           'Pnas (proceedings of the national academy of sciencences of the united states of america',
           'Proceedings of the national academy of sciencences',
           'Pnas', 'PNAS USA','PNAS (Proceedings of the National Academy of Sciencences of the United States of America',
           'Proceedings of the National Academy of Sciencences (PNAS)'
             ]

#correct titles from list to consolodate articles 
for name in pnaslist:
    df['Journal title']= df['Journal title'].replace([name],'PNAS')


In [900]:
# remove whitespace at the begining and end of each jornal tittle 
df['Journal title'] = df['Journal title'].str.strip()

#make all letters lower case in the jornal title column to further condense
df['Journal title'] = df['Journal title'].str.lower()

df['Journal title'] = df['Journal title'].str.replace('jnl', 'journal')

#recapitalize the first letter 
df['Journal title'] = df['Journal title'].str.capitalize()
count=df.groupby('Journal title').nunique()
count.describe()

PMID/PMCID   Publisher  Journal title  Article title  \
count  860.000000  860.000000          860.0      860.00000   
mean     2.234884    1.329070            1.0        2.47093   
std      7.592392    0.843984            0.0        7.95200   
min      0.000000    1.000000            1.0        1.00000   
25%      1.000000    1.000000            1.0        1.00000   
50%      1.000000    1.000000            1.0        1.00000   
75%      2.000000    1.000000            1.0        2.00000   
max    201.000000   13.000000            1.0      210.00000   

       COST (£) charged to Wellcome (inc VAT when charged)  
count                                         860.000000    
mean                                            2.115116    
std                                             6.157007    
min                                             1.000000    
25%                                             1.000000    
50%                                             1.000000    
75%                                             2.000000    
max                                           158.000000

In [901]:
count=df.groupby('Journal title').nunique()
count = count.sort_values('Article title')
#count.describe() 
count.tail(6)
#  get to half of 2126 to be certain


PMID/PMCID  Publisher  Journal title  \
Journal title                                                           
Plos pathogens                           24          2              1   
Plos genetics                            23          5              1   
Nucleic acids research                   25          4              1   
Neuroimage                               30          2              1   
Journal of biological chemistry          63         13              1   
Plos one                                201          5              1   

                                 Article title  \
Journal title                                    
Plos pathogens                              24   
Plos genetics                               24   
Nucleic acids research                      26   
Neuroimage                                  36   
Journal of biological chemistry             65   
Plos one                                   210   

                                 COST (£) charged to Wellcome (inc VAT when charged)  
Journal title                                                                         
Plos pathogens                                                                  22    
Plos genetics                                                                   21    
Nucleic acids research                                                           5    
Neuroimage                                                                      34    
Journal of biological chemistry                                                 62    
Plos one                                                                       158

In [902]:
tot = df.groupby('Journal title').count()
tot = tot.sort_values('Article title')
tot.tail()

PMID/PMCID  Publisher  Article title  \
Journal title                                                           
Plos genetics                            23         24             24   
Nucleic acids research                   25         26             26   
Neuroimage                               30         36             36   
Journal of biological chemistry          64         65             65   
Plos one                                201        211            211   

                                 COST (£) charged to Wellcome (inc VAT when charged)  
Journal title                                                                         
Plos genetics                                                                   24    
Nucleic acids research                                                          26    
Neuroimage                                                                      36    
Journal of biological chemistry                                                 65    
Plos one                                                                       211

I decided to go with the number of unique article titles because it is likely that the ID numbers are reduced due to non existant values.

After data cleaning it appears that the journals with the highest count of articles are 
- 1.Plos One with 211	
- 2.Journal of biological chemistry	with 65
- 3.Neuroimage with 36
- 4.Pnas with 35
- 5.Nucleic acids research with 29



notes on correct answers


Next, calculate the mean, median, and standard deviation of the open-access cost per article for each journal.


In [903]:
#check the datatype for costs 
df['COST (£) charged to Wellcome (inc VAT when charged)'].dtype


dtype('O')

In [904]:
#remove pounds symbol, convert to float, and remove values = 999999.000000


df['COST (£) charged to Wellcome (inc VAT when charged)'] = df['COST (£) charged to Wellcome (inc VAT when charged)'].str.replace('£','')
df['COST (£) charged to Wellcome (inc VAT when charged)'] = df['COST (£) charged to Wellcome (inc VAT when charged)'].str.replace('$','')
df['COST (£) charged to Wellcome (inc VAT when charged)'] = pd.to_numeric(df['COST (£) charged to Wellcome (inc VAT when charged)'],errors='coerce')
df['COST (£) charged to Wellcome (inc VAT when charged)'] = df['COST (£) charged to Wellcome (inc VAT when charged)'].replace(999999.000000, np.NaN)



In [905]:
#create a list of the top jounrals and calculate summary statistics of cost for each jounral

topjornals =[
'Journal of biological chemistry',
'Nucleic acids research',
'Pnas',
'Neuroimage',
'Plos one']

for jnl in topjornals:
    tdf = df.loc[lambda df: df['Journal title'] == jnl, :]
    print(jnl,('\n'), tdf['COST (£) charged to Wellcome (inc VAT when charged)'].describe(),('\n'))


Journal of biological chemistry 
 count      63.000000
mean     1380.104921
std       407.386120
min       265.670000
25%      1144.055000
50%      1287.750000
75%      1583.415000
max      2501.070000
Name: COST (£) charged to Wellcome (inc VAT when charged), dtype: float64 

Nucleic acids research 
 count      26.000000
mean     1149.000000
std       442.940447
min       710.000000
25%       852.000000
50%       852.000000
75%      1704.000000
max      2184.000000
Name: COST (£) charged to Wellcome (inc VAT when charged), dtype: float64 

Pnas 
 count       8.000000
mean      769.781250
std       146.090901
min       625.000000
25%       663.700000
50%       729.700000
75%       836.727500
max      1069.800000
Name: COST (£) charged to Wellcome (inc VAT when charged), dtype: float64 

Neuroimage 
 count      36.000000
mean     2057.318056
std       466.871610
min       987.750000
25%      1762.690000
50%      2289.245000
75%      2396.045000
max      2518.230000
Name: COST (£) charge

In [906]:
#descriptive statistics for cost of all jornals for comparison 

df['COST (£) charged to Wellcome (inc VAT when charged)'].describe()


count      2080.000000
mean       2015.038038
std        6100.422956
min           0.000000
25%        1268.700000
50%        1852.505000
75%        2302.930000
max      201024.000000
Name: COST (£) charged to Wellcome (inc VAT when charged), dtype: float64

For a real bonus round, identify the open access prices paid by subject area.